Retry update_children_docs_by_query with a minimum delay to avoid ES ConflictErrors

albertisfu opened the pull request:
According to [COURTLISTENER-5AM](https://freelawproject.sentry.io/issues/4578580569/), we are still encountering some `ConflictError`s for the `update_children_docs_by_query` method.

I suspect that the issue may be related to the **`refresh_interval`**. The refresh interval could be affecting the **`Update by query`** requests, as these requests also perform a search to retrieve the documents that need to be updated. Consequently, they might be retrieving outdated versions of the documents, leading to **`ConflictErrors`**.

This issue is also discussed here:

- https://discuss.elastic.co/t/elasticsearch-delete-by-query-409-version-conflict/174150/2)
- https://stackoverflow.com/a/66944587

Now that we have extended the **`refresh_interval`** to 5 minutes, as described in issue #3316, this longer refresh rate may affect the UBQ requests when the initial backoff time set in the Celery task is too low. When **`retry_jitter`** is enabled, the initial **`retry_backoff`** becomes a value between 0 and the **`retry_backoff`** setting. For example, if the value is set to 2 seconds, all retries will execute within approximately 32 seconds, which is less than the current 5-minute **`refresh_interval`**. According to the Celery documentation, there is no way to set a minimum value for **`retry_backoff`**.

To address this, an alternative solution implemented in this PR is to remove the auto-retry settings and use a countdown for executing retries. The countdown is set to a random value between 5 and 15 minutes. This approach ensures that retries are performed after the 5-minute **`refresh_interval`** and also prevents simultaneous retries for the same document.

- Additionally, I have removed the outdated Elasticsearch Celery tasks, as none should now be left in the queue.

Let me know what you think.

mlissner commented:
Shoot. I'm not sure I got much of a performance boost by changing the refresh interval. Perhaps we just put it back to the default of 1s?

albertisfu commented:
Yeah, if the performance doesn't improve, we could simply revert to a 1-second delay. However, I still believe that using a random countdown with a minimum and maximum delay is better than using Celery's `retry_backoff` . This is because if the initial backoff is too low (close to zero), and the UBQ request takes longer to complete on retries, we could still encounter `ConflictErrors`. This happens because tasks with conflicts would be retried almost simultaneously.

By using a random delay, for instance, between 1 and 5 minutes, it's more likely that a retried task will finish before another one starts, thereby avoiding conflicts.

We could even consider something like:

`countdown = (number_of_retry * 60) + random(1*60, 5*60)`

What do you think?

mlissner commented:
Your examples feel pretty slow, but also, what about this: https://docs.celeryq.dev/en/stable/userguide/tasks.html#Task.default_retry_delay

albertisfu commented:
Thanks, I tested `default_retry_delay` in combination with `retry_backoff` and `retry_jitter` but in this combination `default_retry_delay` is ignored, instead the initial delay is a random value between 0 and the value set in `retry_backoff`.

`default_retry_delay` works correctly without using `retry_backoff` and `retry_jitter` however using it without a backoff and jitter is likely tasks can be retried at the same time.

What's a reasonable total maximum time for a task to be executed after retries? So I can set a manual retry policy that meets this maximum time and reduces the probability of a task being retried simultaneously.

mlissner commented:
So the balance is if we go to fast, we'll get more conflicts and thus more retries and if we go too slow, the search engine lags behind reality. 

I think conflicts should be relatively rare and that we'll be getting refreshes every second in prod, so maybe we want everything done within 30s?

albertisfu commented:
Ok, according to our conversation, I've updated the countdown for the UBQ task retries as follows:

```
min_delay = 10  # 10 seconds
max_delay = 15  # 15 seconds
countdown = ((retry_count + 1) * min_delay) + randint(
    min_delay, max_delay
)
```
So in this case, the first retry will happen 10 seconds later + a random value between 10 and 15 seconds, and it will continue increasing by around 10 seconds on each retry. So in the worst case, the latest retry (4°) will have a delay of 40 seconds  + a random value between 10 and 15 seconds, this is longer than 30 seconds but I think most of the requests will be retried in the first or two retries. If we still get ConflictErrors we'd need to increase the delays or change the approach to avoid them.

- Reduced the `retry_backoff` in the other ES tasks,  to a value between 0 and 1 minute. Since the other tasks are not getting `ConflictErrors` anymore this makes sense.

- I also added `ignore_result=True` to all the Es indexing tasks. I tested that retries continue working correctly after adding this setting and also that adding this setting to `es_save_document` doesn't affect the chain we use for sending alerts. There are no issues with using `ignore_result=True` in these tasks.

- Removed Old tasks that are not in use anymore and that are not supposed to have an enqueue task by now.

- Lastly, I set the scroll expiration time to 3 minutes. According to [documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-update-by-query.html#_waiting_for_active_shards_2) the default keep-alive time is 5 minutes, we're using 30 slices for each UBQ request, so each request creates 30 scroll context if we get 10 or more UBQ requests in 5 minutes the limit is reached  (500), so reducing it to 3 minutes can help to avoid reaching the limit. We can't reduce it too much since there is a risk of the contexts expiring before the UBQ request finishes. If we continue getting errors related to `max_open_scroll_context` the alternative is to increase the limit in ES, but that could use more resources.
